.github/workflows/e2e-tests.yaml

```
@@ -67,9 +67,52 @@ on:
         default: false
 
 jobs:
-  test-integration:
-    name: Tests
-    runs-on: ${{ inputs.arch == 'arm64' && 'self-hosted-linux-arm64-noble-large' || 'self-hosted-linux-amd64-noble-large' }}
+  # First job: Prepare the snap and find test files
+  prepare:
+    name: Prepare Environment
+    runs-on: ${{ inputs.arch == 'arm64' && 'ubuntu-latest-arm64' || 'ubuntu-latest' }}
+    outputs:
+      tests: ${{ steps.collect-tests.outputs.tests }}
+    steps:
+      - name: Check out code
+        uses: actions/checkout@v4
+      - name: Setup Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.12"
+      - name: Install tox
+        run: sudo apt-get install -y tox
+      - name: Collect tests
+        id: collect-tests
+        run: |
+          cd tests/integration
+          # Split the input tags into an array and pass them as individual args
+          IFS=' ' read -ra TAGS <<< "${{ inputs.test-tags }}"
+          TAGS_ARGS=""
+          for tag in "${TAGS[@]}"; do
+            TAGS_ARGS="$TAGS_ARGS --tags $tag"
+          done
+
+          # Collect test names and convert to JSON array for GitHub Actions
+          TEST_FILES=$(tox -e collect-tests -- --collect-only tests/ $TAGS_ARGS \
+            | grep -E "tests/test_.*\.py" \
+            | sed -E 's/.*::([a-zA-Z0-9_]+).*/\1/' \
+            | sort -u)
+
+          # Convert to JSON array for GitHub Actions
+          JSON_ARRAY=$(echo "$TEST_FILES" | jq -R -s -c 'split("\n") | map(select(length > 0))')
+          echo "tests=$JSON_ARRAY" >> $GITHUB_OUTPUT
+          echo "Found test files: $TEST_FILES"
+
+  # Second job: Run each test in a separate runner
+  run-tests:
+    name: Run ${{ matrix.test }} - ${{ inputs.os }}-${{ inputs.arch }}
+    needs: prepare
+    runs-on: ${{ inputs.arch == 'arm64' && 'ubuntu-latest-arm64' || 'ubuntu-latest' }}
+    strategy:
+      fail-fast: false
```

```suggestion
    strategy:
      max-parallel: 10    
      fail-fast: false
```

this will limit the number of runners consumed to 10, so the jobs don't swamp and eat all the runners at once

---

as discussed offline: We use Github runners not self-hosted ones so we should not need to limit this.